### PR TITLE
Use global pubkey cache for block processing

### LINKED ISSUE #848: Use the beacon chain pubkey cache for block signature verification
## Introduction

Previously we had an issue where the beacon state needs to be serialized with _compressed_ pubkeys but we need to do signature verification with _decompressed_ pubkeys. Decompressing is expensive so we want to avoid doing it every time we verify a signature. Our solution was to add an optional decompressed version to each `PublicKeyBytes` struct as a sort-of cache and just try and make sure that we keep that state in memory between block processing operations (it'll lose the decompressed versions if it goes to disk).

This works pretty well when you always built on top of a block that's in memory, but starts to fall apart when you have 100k validators and blocks are being built of old, non-cached blocks.

## New solution

#841 introduces a `validator_pubkey_cache: TimeoutRwLock<ValidatorPubkeyCache>` which is basically a never-shrinking list of all the _**decompressed**_ validator pubkeys we've seen, in the same order that they appear in the `BeaconState`. It relies heavily upon the idea that some `BeaconState` may have more or less validators, however they are always in the same order.

We can just use this cache during block processing and forget about dealing with the keys in the beacon state.

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

- Resolves #848
- Related to #847

## Proposed Changes

- Uses the beacon chain pubkey cache for block signature verification.
- Changes `get_attesting_indices` to return a `Vec` instead of a `BTreeMap` to reduce heap allocations.
- Makes a (somewhat tangential) change to SSZ enc/dec for BLS objects to avoid unnecessary allocations.

## TODO:

- [x] Persist to disk to cover us in the case that the head has less validators than were previously seen.

## Additional Info

- ~~Blocked on #841~~

### CODE REVIEW

File: eth2/state_processing/src/per_block_processing/block_signature_verifier.rs [hunk 1]
```diff
@@ -46,23 +47,34 @@ impl From<BlockOperationError<AttestationInvalid>> for Error {
...
 /// This allows for optimizations related to batch BLS operations (see the
 /// `Self::verify_entire_block(..)` function).
-pub struct BlockSignatureVerifier<'a, T: EthSpec> {
+pub struct BlockSignatureVerifier<'a, T, F>
+where
+    T: EthSpec,
+    F: Fn(usize) -> Option<Cow<'a, G1Point>> + Clone,
+{
     block: &'a SignedBeaconBlock<T>,
+    get_pubkey: Box<F>,
```
Comment (michaelsproul) [inline eth2/state_processing/src/per_block_processing/block_signature_verifier.rs:56]: This compiles without the `Box`, do we need it?

**Comment by paulhauner** [comment]
I addressed your comments, thank you! The `Option::filter` one was a great idea, I haven't used it before.

I've also commented out some tests failing due to `Eth2TestnetDir::hardcoded` introduced in #912 which broke when applied on top of #959.

File: beacon_node/rest_api/src/validator.rs [hunk 2]
```diff
@@ -189,11 +189,24 @@ fn return_validator_duties<T: BeaconChainTypes>(
...
-                })?
-            {
+            // The `beacon_chain` can return a validator index that does not exist in all states.
+            // Therefore, we must check to ensure that the validator index is valid for our
+            // `state`.
+            let validator_index = if let Some(i) = beacon_chain
+                .validator_index(&validator_pubkey)
+                .map_err(|e| {
+                ApiError::ServerError(format!("Unable to get validator index: {:?}", e))
+            })? {
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/rest_api/src/validator.rs]: Done! Great idea.

File: eth2/state_processing/src/per_block_processing/block_signature_verifier.rs (hunk 1 above)
Comment (paulhauner) [reply eth2/state_processing/src/per_block_processing/block_signature_verifier.rs]: Removed, thank you!

**Pull request merged**